### ISSUE TITLE
ValueError: Input contains NaN, infinity or a value too large for dtype('float32').

### ISSUE THREAD

**Issue by JonathanMathews** [takahikof/ShapeAutoTranscoder#1]
I am trying to run your code using the ModelNet10 dataset as specified in your instructions.

(base) user@vmname:~/ShapeAutoTranscoder$ sudo docker run --rm --runtime=nvidia --gpus all -v $(pwd):/mnt sat:latest nvidia-smi
Sat Aug  5 20:23:55 2023       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.86.10              Driver Version: 535.86.10    CUDA Version: 10.0     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  Tesla T4                       Off | 00000000:00:04.0 Off |                    0 |
| N/A   55C    P0              28W /  70W |      2MiB / 15360MiB |      6%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|  No running processes found                                                           |
+---------------------------------------------------------------------------------------+

I am receiving he following error. on the training step right...

sudo docker run --rm --runtime=nvidia --gpus all -v $(pwd):/mnt gcr.io/colossal-comp-sci/sat:latest ./TrainAndTest.sh
python -u LearnSAT.py ./data/train.shaperep ./data/list_ModelNet10_train.txt ./data/label_ModelNet10_train.txt ./data/test.shaperep ./data/list_ModelNet10_test.txt ./data/label_ModelNet10_test.txt 64 /gpu:0 ./checkpoint
loading files...
Number of training 3D models: 3991
Number of testing 3D models: 908
constructing DNN...
2023-08-05 20:01:47.110831: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2023-08-05 20:01:47.118175: E tensorflow/stream_executor/cuda/cuda_driver.cc:300] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
2023-08-05 20:01:47.118220: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:161] retrieving CUDA diagnostic information for host: 32db2c9276d1
2023-08-05 20:01:47.118233: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:168] hostname: 32db2c9276d1
2023-08-05 20:01:47.118297: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:192] libcuda reported version is: 410.104.0
2023-08-05 20:01:47.118327: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:196] kernel reported version is: 535.86.10
2023-08-05 20:01:47.118337: E tensorflow/stream_executor/cuda/cuda_diagnostics.cc:306] kernel version 535.86.10 does not match DSO version 410.104.0 -- cannot find working devices in this configuration
2023-08-05 20:01:47.120773: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2299995000 Hz
2023-08-05 20:01:47.123001: I tensorflow/compiler/xla/service/service.cc:150] XLA service 0x12b5120 executing computations on platform Host. Devices:
2023-08-05 20:01:47.123038: I tensorflow/compiler/xla/service/service.cc:158]   StreamExecutor device (0): <undefined>, <undefined>
WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/op_def_library.py:263: colocate_with (from tensorflow.python.framework.ops) is deprecated and will be removed in a future version.
Instructions for updating:
Colocations handled automatically by placer.

WARNING: The TensorFlow contrib module will not be included in TensorFlow 2.0.
For more information, please see:
  * https://github.com/tensorflow/community/blob/master/rfcs/20180907-contrib-sunset.md
  * https://github.com/tensorflow/addons
If you depend on functionality not listed there, please file an issue.

set_to_set distance : CHAMFER with L2
WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow/python/ops/math_ops.py:3066: to_int32 (from tensorflow.python.ops.math_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.cast instead.
WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow/python/ops/math_grad.py:102: div (from tensorflow.python.ops.math_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Deprecated in favor of operator or tf.math.divide.
 [*] Reading checkpoint from ./checkpoint
Pretrained parameters do not exist. Start training with randomly initialized parameters.
Number of training epochs : 201
Number of minibatches per epoch : 498
Mini batch size : 8
Epoch: 0
overallloss_train: nan
overallloss_test: nan
extracting features for evaluation...
!!!!!!!!!! error: Inf or Nan has been detected !!!!!!!!!
Combination by vector concatenation:
evaluating retrieval accuracy of testing shapes...
Traceback (most recent call last):
  File "LearnSAT.py", line 90, in <module>
    in_list_filename_test, in_label_filename_test );
  File "utils/model_dnn.py", line 461, in train
    map = self.evaluateRetrievalAccuracy( f_test, in_list_filename_test, in_label_filename_test );
  File "utils/model_dnn.py", line 651, in evaluateRetrievalAccuracy
    nn, map = retrieval.retrieval( feat, in_list_filename, in_label_filename, "COS" );
  File "utils/retrieval.py", line 36, in retrieval
    D = pairwise_distances( feat, metric=metric );
  File "/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py", line 63, in inner_f
    return f(*args, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/sklearn/metrics/pairwise.py", line 1790, in pairwise_distances
    return _parallel_pairwise(X, Y, func, n_jobs, **kwds)
  File "/usr/local/lib/python3.6/dist-packages/sklearn/metrics/pairwise.py", line 1359, in _parallel_pairwise
    return func(X, Y, **kwds)
  File "/usr/local/lib/python3.6/dist-packages/sklearn/metrics/pairwise.py", line 833, in cosine_distances
    S = cosine_similarity(X, Y)
  File "/usr/local/lib/python3.6/dist-packages/sklearn/metrics/pairwise.py", line 1180, in cosine_similarity
    X, Y = check_pairwise_arrays(X, Y)
  File "/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py", line 63, in inner_f
    return f(*args, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/sklearn/metrics/pairwise.py", line 144, in check_pairwise_arrays
    estimator=estimator)
  File "/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py", line 63, in inner_f
    return f(*args, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py", line 721, in check_array
    allow_nan=force_all_finite == 'allow-nan')
  File "/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py", line 106, in _assert_all_finite
    msg_dtype if msg_dtype is not None else X.dtype)
ValueError: Input contains NaN, infinity or a value too large for dtype('float32').

**Comment by JAMKuttan** [comment]
I am having the exact same problem! I have tried multiple versions of TensorFlow and CUDA all with the same outcome.

**Comment by takahikof** [comment]
Thank you for the report.
Your execution results show two TensorFlow-related errors.
* E tensorflow/stream_executor/cuda/cuda_driver.cc:300] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
* E tensorflow/stream_executor/cuda/cuda_diagnostics.cc:306] kernel version 535.86.10 does not match DSO version 410.104.0 -- cannot find working devices in this configuration

It seems that TensorFlow in your environment cannot use the GPU.
First of all, I recommend that you change /gpu:0 to /cpu:0 in the execution command to confirm that a training with CPU does not cause errors.

**Comment by JonathanMathews** [comment]
Thank you for your response. I tried changing /gpu:0 to /cpu:0, but, this did not fix the issue. There are still nan values found and thus computations cannot be completed. The error I am receiving is the same as before, regardless of cpu or gpu:

root@2fb91c582e3d:/mnt# ./TrainAndTest.sh 
python -u LearnSAT.py ./data/train.shaperep ./data/list_ModelNet10_train.txt ./data/label_ModelNet10_train.txt ./data/test.shaperep ./data/list_ModelNet10_test.txt ./data/label_ModelNet10_test.txt 64 /cpu:0 ./checkpoint
loading files...
Number of training 3D models: 3991
Number of testing 3D models: 908
constructing DNN...
WARNING:tensorflow:From LearnSAT.py:79: The name tf.ConfigProto is deprecated. Please use tf.compat.v1.ConfigProto instead.

WARNING:tensorflow:From LearnSAT.py:83: The name tf.Session is deprecated. Please use tf.compat.v1.Session instead.

2023-08-14 17:19:23.760832: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2023-08-14 17:19:24.287518: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2299995000 Hz
2023-08-14 17:19:24.290886: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x44a3330 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2023-08-14 17:19:24.290923: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
2023-08-14 17:19:24.870094: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcuda.so.1
2023-08-14 17:19:25.089667: E tensorflow/stream_executor/cuda/cuda_driver.cc:318] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
2023-08-14 17:19:25.089733: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:163] no NVIDIA GPU device is present: /dev/nvidia0 does not exist
WARNING:tensorflow:From utils/model_dnn.py:69: The name tf.placeholder is deprecated. Please use tf.compat.v1.placeholder instead.

WARNING:tensorflow:From utils/model_dnn.py:576: The name tf.train.exponential_decay is deprecated. Please use tf.compat.v1.train.exponential_decay instead.

WARNING:tensorflow:From utils/ops_dnn.py:475: The name tf.matrix_band_part is deprecated. Please use tf.linalg.band_part instead.

WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow_core/python/ops/array_ops.py:1475: where (from tensorflow.python.ops.array_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where
WARNING:tensorflow:From utils/ops_dnn.py:71: The name tf.variable_scope is deprecated. Please use tf.compat.v1.variable_scope instead.

WARNING:tensorflow:From utils/ops_dnn.py:72: The name tf.get_variable is deprecated. Please use tf.compat.v1.get_variable instead.

WARNING:tensorflow:
The TensorFlow contrib module will not be included in TensorFlow 2.0.
For more information, please see:
  * https://github.com/tensorflow/community/blob/master/rfcs/20180907-contrib-sunset.md
  * https://github.com/tensorflow/addons
  * https://github.com/tensorflow/io (for I/O related ops)
If you depend on functionality not listed there, please file an issue.

WARNING:tensorflow:From utils/ops_dnn.py:136: The name tf.segment_max is deprecated. Please use tf.math.segment_max instead.

WARNING:tensorflow:From utils/ops_dnn.py:41: The name tf.nn.max_pool is deprecated. Please use tf.nn.max_pool2d instead.

WARNING:tensorflow:From utils/model_dnn.py:308: The name tf.random_shuffle is deprecated. Please use tf.random.shuffle instead.

set_to_set distance : CHAMFER with L2
WARNING:tensorflow:From utils/model_dnn.py:107: The name tf.log is deprecated. Please use tf.math.log instead.

WARNING:tensorflow:From utils/model_dnn.py:120: The name tf.train.Saver is deprecated. Please use tf.compat.v1.train.Saver instead.

WARNING:tensorflow:From utils/model_dnn.py:360: The name tf.train.AdamOptimizer is deprecated. Please use tf.compat.v1.train.AdamOptimizer instead.

WARNING:tensorflow:From utils/model_dnn.py:363: The name tf.global_variables_initializer is deprecated. Please use tf.compat.v1.global_variables_initializer instead.

 [*] Reading checkpoint from ./checkpoint
Pretrained parameters do not exist. Start training with randomly initialized parameters.
Number of training epochs : 201
Number of minibatches per epoch : 498
Mini batch size : 8
Epoch: 0
overallloss_train: nan
overallloss_test: nan
extracting features for evaluation...
!!!!!!!!!! error: Inf or Nan has been detected !!!!!!!!!
Combination by vector concatenation:
evaluating retrieval accuracy of testing shapes...
Traceback (most recent call last):
  File "LearnSAT.py", line 90, in <module>
    in_list_filename_test, in_label_filename_test );
  File "utils/model_dnn.py", line 461, in train
    map = self.evaluateRetrievalAccuracy( f_test, in_list_filename_test, in_label_filename_test );
  File "utils/model_dnn.py", line 651, in evaluateRetrievalAccuracy
    nn, map = retrieval.retrieval( feat, in_list_filename, in_label_filename, "COS" );
  File "utils/retrieval.py", line 36, in retrieval
    D = pairwise_distances( feat, metric=metric );
  File "/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py", line 63, in inner_f
    return f(*args, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/sklearn/metrics/pairwise.py", line 1790, in pairwise_distances
    return _parallel_pairwise(X, Y, func, n_jobs, **kwds)
  File "/usr/local/lib/python3.6/dist-packages/sklearn/metrics/pairwise.py", line 1359, in _parallel_pairwise
    return func(X, Y, **kwds)
  File "/usr/local/lib/python3.6/dist-packages/sklearn/metrics/pairwise.py", line 833, in cosine_distances
    S = cosine_similarity(X, Y)
  File "/usr/local/lib/python3.6/dist-packages/sklearn/metrics/pairwise.py", line 1180, in cosine_similarity
    X, Y = check_pairwise_arrays(X, Y)
  File "/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py", line 63, in inner_f
    return f(*args, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/sklearn/metrics/pairwise.py", line 144, in check_pairwise_arrays
    estimator=estimator)
  File "/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py", line 63, in inner_f
    return f(*args, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py", line 721, in check_array
    allow_nan=force_all_finite == 'allow-nan')
  File "/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py", line 106, in _assert_all_finite
    msg_dtype if msg_dtype is not None else X.dtype)
ValueError: Input contains NaN, infinity or a value too large for dtype('float32').

**Comment by takahikof** [comment]
In that case, I suspect the cause of the error is in the data you created. Below are the .shaperep files that I created in my environment. I have confirmed that running the code with these .shaperep files does not cause an error. 

test.shaperep
https://drive.google.com/file/d/1YIS8QfmSSdlevp7oGaUMgqNpFd9_dfwo/view?usp=drive_link

train.shaperep
https://drive.google.com/file/d/1imniyayR5xvwkDJwVocr6TjYZaTKFd7z/view?usp=drive_link

If the NaN error disappears with my .shaperep files, you may need to revise your environment such as versions of python libraries.  (However, it is difficult for me to give you further advice about linux environment or python library compatibility since I'm not familiar with them. I apologize in advance).

**Comment by JonathanMathews** [comment]
I found the issue, and it had to do with my environment libraries. The issue occured with tensorflow. Though I have tensorflow-gpu 1.13 installed, the tensorflow.contrib module is fully deprecated even in the tensorflow 1.x distribution. I had to replace the tensorflow.contrib functions in the source code with the corresponding functions from tensorflow 2.x (since some of the important functions from tensorflow.contrib were encompassed in the official tensorflow module in version 2.x). Now the training is proceeding smoothly with no issues. Thank you for your help!
